pkg/server/server.go

```diff
@@ -222,15 +242,90 @@ func (s *Server) Run(ctx context.Context) error {
 		KeyFile:       s.cfg.EtcdClientInfo.KeyFile,
 		TrustedCAFile: s.cfg.EtcdClientInfo.TrustedCAFile,
 	}
+
+	// TODO(ncdc): I thought I was going to need this, but it turns out this breaks the CRD controllers because they
+	// try to issue Update() calls using the * client, which ends up with the cluster name being set to the default
+	// admin cluster in handler.go. This means the update calls are likely going against the wrong logical cluster.
+	// serverOptions.APIExtensionsNewClientFunc = func(config *rest.Config) (apiextensionsclient.Interface, error) {
+	// 	const crossCluster = "*"
+	// 	clusterClient, err := apiextensionsclient.NewClusterForConfig(config)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// 	client := clusterClient.Cluster(crossCluster)
+	// 	return client, nil
+	// }
+
+	loopbackClientCert, loopbackClientCertKey, err := serverOptions.SecureServing.NewLoopbackClientCert()
+	if err != nil {
+		return err
+	}
+	serverOptions.SecureServing.LoopbackCertPEM = loopbackClientCert
+	serverOptions.SecureServing.LoopbackKeyPEM = loopbackClientCertKey
+	serverOptions.SecureServing.LoopbackToken = uuid.New().String()
+	loopbackClientConfig, err := serverOptions.SecureServing.NewLoopbackClientConfig(serverOptions.SecureServing.LoopbackToken, serverOptions.SecureServing.LoopbackCertPEM)
+	if err != nil {
+		return err
+	}
+
+	// FIXME: switch to a single set of shared informers
+	const crossCluster = "*"
+	kcpClusterClient, err := kcpclient.NewClusterForConfig(loopbackClientConfig)
+	if err != nil {
+		return err
+	}
+	kcpClient := kcpClusterClient.Cluster(crossCluster)
+
+	kcpSharedInformerFactory := kcpexternalversions.NewSharedInformerFactoryWithOptions(kcpClient, resyncPeriod)
+	s.AddPostStartHook("FIXME-start-informers-for-crd-getter", func(context genericapiserver.PostStartHookContext) error {
+		kcpSharedInformerFactory.Start(context.StopCh)
+		return nil
+	})
+	// FIXME: (end) switch to a single set of shared informers
+
+	workspaceLister := kcpSharedInformerFactory.Tenancy().V1alpha1().Workspaces().Lister()
+
+	serverOptions.APIExtensionsNewSharedInformerFactoryFunc = func(client apiextensionsclient.Interface, resyncPeriod time.Duration) crdexternalversions.SharedInformerFactory {
+		f := apiextensionsinformers.NewSharedInformerFactory(client, resyncPeriod)
+		return &kcpAPIExtensionsSharedInformerFactory{
+			SharedInformerFactory: f,
+			workspaceLister:       workspaceLister,
+		}
+	}
```

Comment:
how do you make sure now that the apiextensions controllers don't follow the inheritence implemented here in that informer?

Reply:
Right now via 2 different ways of "cheating":

1. Instead of specifying our own new client func (which would create a `*` client), I use what we had previously - the client is "enabled for multi-cluster", which means it uses our custom multi-cluster round tripper, which attempts to set the cluster on the request based on the ObjectMeta from the request body in a write operation.

2. I didn't undo a previously-written hack in the NamingConditionController that makes it lcluster-aware (it excludes CRDs from other lclusters when listing & looping).

We probably need a proper plan here...

Reply:
Reconciling CRDs in all lclusters is fine and correct. Just the inheritence should not be applied here. I understand from (1) that we don't do that :+1: